Add "Installing the Operator Framework"

adellape opened the pull request:
Preview:

http://file.rdu.redhat.com/~adellape/092618/olm_311/install_config/installing-operator-framework.html

adellape commented on modules/installing-olm-using-ansible.adoc:
```diff
@@ -0,0 +1,59 @@
+
+- An existing {product-title} 3.11 cluster
+- Access to the cluster using an account with `cluster-admin` permissions
+- Ansible playbooks provided by the latest `openshift-ansible` installer
```
I only put it this way in hopes that they would have the latest fixes.

robszumski replied:
If we can do a `{product-version}` or some thing that can help I guess, but I am not really worried about this.

robszumski commented on modules/installing-olm-using-ansible.adoc:
```diff
@@ -0,0 +1,59 @@
+.Prerequisites
+
+- An existing {product-title} 3.11 cluster
+- Access to the cluster using an account with `cluster-admin` permissions
```
No, this doc is only about installing it, which is done by admins. I agree that having a general usage doc is helpful.

jianzhangbjz replied:
Thanks! @robszumski There is the description section about the usage of the OLM in this doc too. But, because of the usage for the general users is not very clear in 3.11, I suggest that we highlight the point which the operations of the OLM(including create operators via the OLM actions, etc...) only for the `cluster-admin` users. What do you think?

adellape replied:
@jianzhangbjz @robszumski @ecordell How about a [NOTE] box in "Launching Your First Operator" that says something like:

"During this Technology Preview phase, focus is placed on cluster administrator usage. The following procedure can currently only be run by a user with cluster-admin permissions, however in upcoming releases, developer users with common permissions will be able to perform certain OLM actions themselves."

Also added a truncated version to "What's in the Technology Preview?" earlier on:

"During this Technology Preview phase, focus is placed on cluster administrator usage. In upcoming releases, developers users with common permissions will be able to perform certain OLM actions themselves."

Better way to say this?

robszumski replied:
We have the functionality for folks to delegate access today. Various sentences that can be strung together however you'd like:

"Installing an Operator within a namespace is done through a Subscription object, which the cluster administrator can create and manage across the entire cluster. Users with a Role that includes access to Subscriptions (eg, `admin` role applied within a Project) will also be able to manage subscriptions for their project."

"Within a Project, all users with the `edit` role will be able to create, manage and delete application instances managed by Operators that have been installed in the Project, in a self-service manner, just like a cloud service."

"Further granularity can be established with a customized Role to grant or prevent users the ability to manage certain Operator instance types."

jianzhangbjz replied:
>  Users with a Role that includes access to Subscriptions (eg, admin role applied within a Project) will also be able to manage subscriptions for their project."

@robszumski Yes, you're right. But, this user cannot access the `CatalogSource` to init their `Subscriptions`. They will get the below errors unless the `cluster-admin` users grant the permission: `# oc policy add-role-to-user view jiazha2 -n operator-lifecycle-manager`
```
catalogsources.operators.coreos.com is forbidden: User "jiazha2" cannot list catalogsources.operators.coreos.com in the namespace "operator-lifecycle-manager": no RBAC policy matched
``` 
And, they cannot create the Cluster against their Operator either since no approprivate permission.  For example, if they want to create the etcd-cluster, they need to get the permission first:
```
# oc policy add-role-to-user etcdoperator.v0.9.2-role-9rswg jiazha2 --role-namespace=jiazha
```

robszumski commented on modules/installing-olm-using-ansible.adoc:
```diff
@@ -0,0 +1,59 @@
+containers:
++
+----
+openshift_additional_registry_credentials=[{'host':'registry.connect.redhat.com','user':'<your_user_name>','password':'<your_password>','image_name':'mongodb/enterprise-operator:0.3.2'}]
```
This landed for 3.11 and should lay down credentials for the kubelet, so it should be used for any container image coming from registry.connect.redhat.com. It doesn't necessarily need to be called by that playbook as far as I understand it.

jianzhangbjz replied:
Yes, but we should make the `openshift_additional_registry_credentials` work, right? So, run which playbook could make it work, we should document this in here.

jianzhangbjz replied:
@ecordell Sorry for the late to reply. Yeah, I mean we just should point out running which playbook could install the `openshift_additional_registry_credentials`. We don't need to call this in the OLM playbook.

adellape replied:
Added a step to run `playbooks/updates/registry_auth.yml` before `playbooks/olm/config.yml`.

jianzhangbjz commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,119 @@
+  password: password
+----
+
+. Create a new `CouchbaseCluster` object, which is a new object that the Operator
```
@adellape Maybe we should clarify this in here.

jianzhangbjz replied:
@ecordell Actually, no. The users still need to the refresh operation after they click the `create cluster` button.

adellape replied:
Added a [TIP] box to suggest refreshing as a workaround, and cited the BZ.

jianzhangbjz replied:
That's good! Thanks!

adellape commented on modules/installing-olm-using-ansible.adoc:
```diff
@@ -0,0 +1,60 @@
+====
+Alternatively, the Technology Preview Operator Framework can be  installed
+during initial cluster installation. See
+xref:../install/configuring_inventory_file.adoc#install-config-configuring-inventory-file[Configuring Your Inventory File] for separate instructions.
```
@ecordell Yeah, sorry, have not added that yet.

adellape commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Document: Operator Lifecycle Manager installed in openshift-ansible](https://trello.com/c/AD7HEdax/979-document-operator-lifecycle-manager-installed-in-openshift-ansible)

adellape commented:
@ecordell @jianzhangbjz Added the procedure for installing OLM during initial cluster installing to the "Installing Clusters" guide, PTAL:

http://file.rdu.redhat.com/~adellape/091818/olm_311/install/configuring_inventory_file.html#configuring-the-operator-lifecycle-manager

So now the link in the "NOTE" box here actually goes somewhere:

http://file.rdu.redhat.com/~adellape/091818/olm_311/operators/installing-operator-framework.html#installing-olm-using-ansible_installing-operator-framework

jianzhangbjz commented on install/configuring_inventory_file.adoc:
```diff
@@ -2018,3 +2018,58 @@ following variables can be used to customize the cluster console.
+
+Installing the Technology Preview
+link:https://coreos.com/blog/introducing-operator-framework[Operator Framework]
+is handled via an Operator. You can install the Operator Lifecycle Manager (OLM)
```
Is the installing of the OLM handled by the Operator? @ecordell

adellape replied:
I wondered about this line as well; came from https://docs.google.com/document/d/1Rh3bSMw0CPnZQBws5AnhBBcesL-Mcs8b9xvga-icdG4/edit#heading=h.2abu0i1tlsjd. Can drop it.

adellape replied:
I've dropped it.

adellape commented:
@jianzhangbjz I've addressed comments, as well as moved the topic under the "Configuring Clusters" guide, instead of its own "Operators" guide. PTAL.

jianzhangbjz commented:
@adellape Thanks! I'm sorry, I could not find the latest preview link, can you help post it in here?

jianzhangbjz commented:
> openshift_additional_registry_credentials=[{'host':'registry.connect.redhat.com','user':'<your_user_name>','password':'<your_password>','image_name':'mongodb/enterprise-operator:0.3.2'}]

Actually, the `0.3.2` version no longer exist, but the `latest` version did exist. See below:
```
[root@qe-jiazha-node-1 ~]#  docker pull registry.connect.redhat.com/mongodb/enterprise-operator:0.3.2
Trying to pull repository registry.connect.redhat.com/mongodb/enterprise-operator ... 
Pulling repository registry.connect.redhat.com/mongodb/enterprise-operator
Error: image mongodb/enterprise-operator:0.3.2 not found

[root@qe-jiazha-node-1 ~]#  docker pull registry.connect.redhat.com/mongodb/enterprise-operator
Using default tag: latest
Trying to pull repository registry.connect.redhat.com/mongodb/enterprise-operator ... 
latest: Pulling from registry.connect.redhat.com/mongodb/enterprise-operator
313bb043666e: Pull complete 
514c0a80c7bf: Pull complete 
baacce39bfd8: Pull complete 
a0c0cc1fca53: Pull complete 
Digest: sha256:2856093081eec1e20369f55e29e8b6aa39e2ee34cd3dd92f3181711dc547ea8f
Status: Downloaded newer image for registry.connect.redhat.com/mongodb/enterprise-operator:latest
```
@ecordell I think this will lead to the MongoDB operator installing fail, so I change their image in their Deployment, but got errors below, how to solve it?
```
[root@qe-jiazha-master-etcd-1 ~]# oc edit deployment mongodb-enterprise-operator
A copy of your changes has been stored to "/tmp/oc-edit-rdp4l.yaml"
error: map: map[] does not contain declared merge key: name
```

@robszumski @ecordell @adellape  For the `playbooks/updates/registry_auth.yml` running, I got the below errors. I didn't know why it failed, I can get the image sotred in the `registry.connect.redhat.com` correctly.
```
$ export REG_AUTH_PASSWORD=NYFQrLg221St9aJtSAmFwp4TTqvsFm34yGCG75LLMGE
$ export REG_AUTH_USER=aos-qe-pull36
$ ansible-playbook -i qe-inventory-host-file playbooks/updates/registry_auth.yml
...
TASK [container_runtime : Create for any additional registries] ***************************************************************************************************************************************************
Wednesday 26 September 2018  13:23:59 +0800 (0:00:02.985)       0:01:55.488 *** 
FAILED - RETRYING: Create for any additional registries (3 retries left).
FAILED - RETRYING: Create for any additional registries (3 retries left).
FAILED - RETRYING: Create for any additional registries (3 retries left).
FAILED - RETRYING: Create for any additional registries (2 retries left).
FAILED - RETRYING: Create for any additional registries (2 retries left).
FAILED - RETRYING: Create for any additional registries (2 retries left).
FAILED - RETRYING: Create for any additional registries (1 retries left).
FAILED - RETRYING: Create for any additional registries (1 retries left).
FAILED - RETRYING: Create for any additional registries (1 retries left).
failed: [qe-jiazha-node-registry-router-1.0926-wma.qe.rhcloud.com] (item={'host': u'registry.connect.redhat.com', 'password': u'xxx', 'user': u'xxx', 'image_name': u'mongodb/enterprise-operator'}) => {"attempts": 3, "changed": false, "item": {"host": "registry.connect.redhat.com", "image_name": "mongodb/enterprise-operator", "password": "xxx", "user": "xxx"}, "msg": "time=\"2018-09-26T05:24:26Z\" level=fatal msg=\"unauthorized: authentication required\" \n", "state": "unknown"}
```

adellape commented:
> @adellape Thanks! I'm sorry, I could not find the latest preview link, can you help post it in here?

@jianzhangbjz Sorry, the preview link is the same as before:

http://file.rdu.redhat.com/~adellape/091818/olm_311/operators/installing-operator-framework.html

jianzhangbjz commented:
@adellape It works now! Thanks a lot! This doc looks good to me.
```
INSTALLER STATUS **************************************************************************************************************************************************************************************************
Initialization  : Complete (0:01:56)
Thursday 27 September 2018  10:07:21 +0800 (0:00:21.922)       0:04:16.431 **** 
=============================================================================== 
Gathering Facts ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 32.75s
Gathering Facts ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 31.08s
openshift_control_plane : verify API server --------------------------------------------------------------------------------------------------------------------------------------------------------------- 21.92s
container_runtime : Create for any additional registries -------------------------------------------------------------------------------------------------------------------------------------------------- 21.92s
Run variable sanity checks -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.75s
Wait for node to be ready --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 16.70s
Gather Cluster facts --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 7.79s
openshift_control_plane : restart master ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 5.95s
restart node ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 5.63s
openshift_node : Create credentials for any additional registries ------------------------------------------------------------------------------------------------------------------------------------------ 4.47s
openshift_control_plane : Create credentials for any additional registries --------------------------------------------------------------------------------------------------------------------------------- 4.38s
container_runtime : Create credentials for oreg_url -------------------------------------------------------------------------------------------------------------------------------------------------------- 4.19s
Initialize openshift.node.sdn_mtu -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 4.17s
openshift_control_plane : Create credentials for oreg_url -------------------------------------------------------------------------------------------------------------------------------------------------- 2.97s
openshift_node : Create credentials for registry auth ------------------------------------------------------------------------------------------------------------------------------------------------------ 2.74s
Detecting Operating System from ostree_booted -------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.75s
get openshift_current_version ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 1.74s
Determine Atomic Host Docker Version ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.71s
openshift_node : Check for credentials file for registry auth ---------------------------------------------------------------------------------------------------------------------------------------------- 1.66s
stat ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.65s
```

>  We don't want to reference latest because the CSV might not be compatible with that release.

@robszumski Yes, I agree. I think we should build the `0.3.2` tagged image and push it to the `registry.connect.redhat.com` registry ASAP, hence we don't need to change our code. Do you think so?
If yes, can you help do this? I don't know who charge of this kind of work. 
But, for this doc, I think the `latest` tagged is OK since it just a testing image used to test the additional registry credentials. What do you think?

adellape commented:
@jianzhangbjz @robszumski @ecordell I've changed the `latest` tag back to `0.3.2` for `mongodb/enterprise-operator`, as I have seen that the tag is back up on the registry as of 12 hours ago:

https://access.redhat.com/containers/?tab=tags#/registry.connect.redhat.com/mongodb/enterprise-operator

Also PTAL again at the "Launching your first Operator" procedure, as I've updated it with the following:

1. Now includes some [NOTE] boxes along the way explaining what roles are required to perform which OLM actions, as well as the `oc policy` commands for how to add them as a `cluster-admin` to an arbitrary user. Note that I found that even if a user had `admin` permissions in a project, they weren't able to actually view the created Instances from the CSV screen (for example, the Couchbase cluster itself after successfully creating it) unless they also had `view` explicitly, so I've mentioned ensuring both are present.
2. Removed the two [NOTE] boxes about "During this Technology Preview phase, focus is placed on cluster administrator usage."
3. Updated screenshots.
4. Re-flowed some of the steps based on my latest testing.

Latest preview: http://file.rdu.redhat.com/~adellape/092618/olm_311/install_config/installing-operator-framework.html#launching-your-first-operator_installing-operator-framework

adellape commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,201 @@
+$ oc policy add-role-to-user view <user> -n operator-lifecycle-manager
+----
+
+This experience will be simplified in future releases of the OLM.
```
@robszumski Keep/drop this?

robszumski replied:
yeah that is fine

robszumski replied:
...to keep it

jianzhangbjz commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,204 @@
+Then create the Subscription using the CLI:
++
+----
+$ oc create -f couchbase-subscription.yaml -n couchbase-test
```
no need the ` -n couchbase-test`

adellape replied:
Oh yeah. Thanks!

jianzhangbjz commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,204 @@
+- For the CLI method, save the secret definition to a YAML file (for example, *_couchbase-secret.yaml_*) and use the CLI to create it in the desired project:
++
+----
+$ oc create -f couchbase-secret.yaml -n couchbase-test
```
no need `-n couchbase-test`

jianzhangbjz commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,204 @@
+
+. Create the new Couchbase cluster.
++
+[NOTE]
```
The line 127 ~ 139 is unnecessary since we already grant the `admin` role to the general user. So, I suggest removing them.

adellape replied:
@jianzhangbjz Are you referring to the `oc policy` steps from lines 32 ~ 40? I phrased that [NOTE] box as optional, if they (optionally) wanted another user that could create Subscriptions. But this line 127 ~ 139 is just for if they (optionally) wanted another user that could manage Instances (but not have the fuller Subscription-level permissions).

jianzhangbjz replied:
OK, I see, thanks!

ahardin-rh commented on modules/launching-your-first-operator.adoc:
```diff
@@ -0,0 +1,204 @@
+<1> Ensure the `namespace` field in the `metadata` section is set to the desired
+project.
++
+Then create the Subscription using the CLI:
```
Then, create

ahardin-rh commented:
@adellape Just one tiny comment from me. Otherwise, looks great!

adellape commented:
/cherry-pick enterprise-3.11

adellape commented:
/cherrypick enterprise-3.11
